### ISSUE TITLE
FR: Kubernetes Operator Should Manage Sidecar Containers to Access Tailnet

### ISSUE THREAD

**Issue by sspreitzer** [tailscale/tailscale#7606]
### What are you trying to do?

Automatically adding tailscale sidecar containers to kubernetes pods.

### How should we solve this?

If a kubernetes namespace is flagged for tailscale, the tailscale kubernetes operator should automatically add tailscale sidecar containers to all pods in the namespace. Eg. with a namespace annotation `tailscale.com/inject-sidecar: "true»`
Also the Kubernetes Operator could be instructed to watch all namespaces, specific namespaces or no namespaces. 

### What is the impact of not solving this?

Tailscale sidecar creation will stay a manual process.

**Comment by maisem** [comment]
Can you provide some details on the use case?

**Comment by sspreitzer** [comment]
> Can you provide some details on the use case?

The use case is to connect workload/applications/pods in a kubernetes namespace automatically to a Tailnet. This functionality is almost equivalent to a kubernetes service mesh like Istio or Linkerd. 

However it can be argued that it would be better to use tailscale to handle the interconnect between kubernetes service mesh gateways.

Maybe this feature request needs to be discussed and voted on.

**Comment by uhthomas** [comment]
This would be really nice. I would really want this for allowing pods to use exit nodes seamlessly, though being able to access other tailscale nodes would also be nice.

**Comment by GMartinez-Sisti** [comment]
Also looking into this feature. We are already using the k8s-operator to expose services into the tailnet, but we also need some workloads to be part of the tailnet so they can access resources back.

**Comment by knyar** [comment]
> we also need some workloads to be part of the tailnet so they can access resources back.

I think you should be able to use [cluster egress](https://tailscale.com/kb/1236/kubernetes-operator/#cluster-egress) for that.

**Comment by GMartinez-Sisti** [comment]
> > we also need some workloads to be part of the tailnet so they can access resources back.
> 
> I think you should be able to use [cluster egress](https://tailscale.com/kb/1236/kubernetes-operator/#cluster-egress) for that.

Thanks, but this doesn’t work for us. Requiring you to know the tailnet ip to create the annotation in advance it’s not compatible with dynamic infrastructure. I’m even surprised this is an option, should use dns or something else.

Made it work with a custom dns config for the pod along with a tailscale-client sidecar plus custom-dns-server sidecar that knows if it needs to query the tailscale-client sidecar or core-dns to resolve to the correct IPs. Can’t share the details, but I can say that you probably don’t want to develop your custom dns server!

**Comment by evilhamsterman** [comment]
I agree with @GMartinez-Sisti it would be nice to have the tailscale operator provide a mutator to install sidecars. The ClusterEgress is on option but it is clunky. You have to specify either the TailNet name or the IP which requires knowing them in advance and then the name that actually is used in the cluster is not the normal name on the tailnet instead you get something like ts-<servicename>-<randomnumber> it just doesn't all link up easily. 

I just hit a use case, I'm doing a POC using nothing but Tailscale to access resources in our cluster and trying to get Coder working just doesn't work for a couple reasons. Coder can't specify separate urls so if I use the operator to present Coder on the Tailnet the server yells about not being able to connect to it's own url. Also the workspace pods aren't able to start because they can't reach the Coder server over the Tailnet. I can put a sidecar in the deployment for the workspace pods, but that requires setting up more oauth keys and stuff.

The ironic thing is Coder uses Tailscale under the hood 😄 

Using Tailscale to replace a service mesh like Istio would be great, it's already almost there with ACLs to control connections, it just needs to be able to inject itself as a sidecar like Istio.

**Comment by rowanmoul** [comment]
I also have a use-case that would also benefit from this functionality. I am using `external-dns` in Cluster A to create DNS records in our own DNS provider that resolve to tailnet IPs (tailscale operator proxies defined in Cluster A).
In Cluster B I want certain pods to be able to use these DNS records to reach tailnet IPs, but this doesn't work unless those pods have a tailscale sidecar to route requests to those IPs.
An alternative would be a more generic tailnet egress proxy deployed in the cluster that will route requests from any pod (network policies permitting) to any tailnet IP, but that would probably require static routes in the cluster or it's underlying network.
Yes I could do this with MagicDNS names and egress proxies, but as others have said, that solution is very clunky and requires me to manually read a nameserver IP from a resource deployed by the operator, and then enter that IP into the core-dns config on my cluster, which is not dynamic or automated.

**Comment by DidZ0** [comment]
Got a usecase today which would've been cool if it was handled by the operator,
I tried to use a mullvad exit node on a pod running a torrent client. I had to use a sidecar container to make it work

**Comment by repository** [comment]
would also love to see this implemented. i'm using tailscale to expose traefik in my cluster (which then handles further routing to services) but i'm losing the client IPs and only getting the tailscale pod IP.

`Ingress` wouldn't work either for me since i need to be able to support http and serve my own certs for other domains

**Comment by tidux** [comment]
This should be considered a bug fix rather than a feature request given how spectacularly running `tailscaled` on a Kubernetes node breaks `core-dns` resolution.

**Comment by spongenee** [comment]
this will solve the usecase mentioned here too https://github.com/tailscale/tailscale/issues/16272
for pods to access dns behind subnet routers

**Comment by fmei-dm** [comment]
would also love to see that.
